<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.5.0 -->
<title>Tutorial | DESC</title>
<meta name="generator" content="Jekyll v3.8.5" />
<meta property="og:title" content="Tutorial" />
<meta property="og:locale" content="en_US" />
<link rel="canonical" href="http://localhost:4000/desc/tutorial.html" />
<meta property="og:url" content="http://localhost:4000/desc/tutorial.html" />
<meta property="og:site_name" content="DESC" />
<script type="application/ld+json">
{"@type":"WebPage","url":"http://localhost:4000/desc/tutorial.html","headline":"Tutorial","@context":"http://schema.org"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/desc/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/desc/feed.xml" title="DESC" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/desc/">DESC</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/desc/">Home</a><a class="page-link" href="/desc/installation.html">Installation</a><a class="page-link" href="/desc/tutorial.html">Tutorial</a><a class="page-link" href="/desc/contact.html">Contact</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <p>In this tutorial, we will perform an entire desc analysis with a dataset of Peripheral Blood Mononuclear Cells (PBMC). In the dataset, the expression levels of 2,700 cells were sequenced using the Illumina NextSeq 500. The data is freely available from 10X Genomics and the raw can be downloaded <a href="https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">here</a>.</p>

<p><br /></p>

<h1 id="0-import-python-modules">0 Import python modules</h1>

<p>In the tutorial, we will need <code class="highlighter-rouge">desc</code> module for clustering analysis and <code class="highlighter-rouge">scanpy</code> for data preprocessing.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">desc</span>
<span class="kn">import</span> <span class="nn">scanpy.api</span> <span class="k">as</span> <span class="n">sc</span>
</code></pre></div></div>

<h1 id="1-import-data">1 Import data</h1>

<p>The current version of desc works with a <a href="https://anndata.readthedocs.io/en/latest/anndata.AnnData.html">AnnData</a> object. AnnData stores a data matrix <code class="highlighter-rouge">.X</code> together with annotations of observations <code class="highlighter-rouge">.obs</code>, variables <code class="highlighter-rouge">.var</code> and unstructured annotations <code class="highlighter-rouge">.uns</code>. The <strong>desc</strong> package provides 3 ways to prepare an AnnData object for the following analysis:</p>

<h2 id="11-start-from-a-10x-dataset">1.1 Start from a 10X dataset</h2>

<p>The desc package provides a function to load the 10X dataset directly. Download the <a href="https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">data</a> and unzip it. Then put everything in <em>filtered_gene_bc_matrices/hg19/</em> to <em>data/pbmc/</em>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load the 10X data by providing the path to data
</span><span class="n">adata</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">read_10X</span><span class="p">(</span><span class="n">data_path</span><span class="o">=</span><span class="s">'data/pbmc'</span><span class="p">)</span>
</code></pre></div></div>

<h2 id="12-start-from-mtx-and-csv-files">1.2 Start from .mtx and .csv files</h2>

<p>When the expression data doesn’t follow the standard 10X dataset format, we can manually import the data as follows:<br />
Read the expression matrix from .mtx file.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">utilities</span><span class="o">.</span><span class="n">read_mtx</span><span class="p">(</span><span class="s">'data/pbmc/matrix.mtx'</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
</code></pre></div></div>

<p>Read the .csv file for gene annotations. Make sure the gene names are unique.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">genes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'data/pbmc/genes.tsv'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s">'gene_ids'</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s">'gene_symbols'</span><span class="p">]</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="c1"># Make sure the gene names are unique
</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">genes</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">var_names</span><span class="o">.</span><span class="n">is_unique</span>
<span class="n">var_names</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">utilities</span><span class="o">.</span><span class="n">make_index_unique</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Index</span><span class="p">(</span><span class="n">var_names</span><span class="p">))</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">var_names</span>
</code></pre></div></div>

<p>Read the .csv file for cell annotations.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">cells</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s">'data/pbmc/barcodes.tsv'</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'barcode'</span><span class="p">]</span> <span class="o">=</span> <span class="n">cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span> <span class="o">=</span> <span class="n">cells</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</code></pre></div></div>

<h2 id="13-start-from-a-h5ad-file">1.3 Start from a .h5ad file</h2>
<p>Or load a previously saved AnnData object:</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s">'data/pbmc.h5ad'</span><span class="p">)</span>
</code></pre></div></div>

<p>For more ways to create AnnData objects, please check <a href="https://anndata.readthedocs.io/en/latest/api.html#reading">here</a></p>

<p><br /></p>

<h1 id="2-preprocessing">2 Preprocessing</h1>

<p>The standard scRNA data preprocessing workflow may include filtration of cells/genes, normalization, scaling and selection of highly variables genes. In this tutorial, we use <code class="highlighter-rouge">scanpy</code> to preprocess the data. Note that among the preprocessing steps, filtration of cells/genes and selecting highly variables genes are optional while <strong>normalization and scaling are strictly required before the desc analysis</strong>.</p>

<p>Following steps show a topical pre-processing procedure for the human blood cells data:</p>

<h2 id="21-filtering-cells-and-genes">2.1 Filtering cells and genes</h2>

<p>Typing <code class="highlighter-rouge">adata</code> in the python console shows that the original datasets contain expressions for 2700 gene across 32738 cells.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;&gt;&gt; adata
AnnData object with n_obs × n_vars = 2700 × 32738 
    obs: 'barcode', 'n_genes'
    var: 'gene_ids', 'gene_symbols', 'n_cells'
</code></pre></div></div>

<p>Let us remove cells with very few expressed genes, as well as genes are lowly expressed, using the function provided by <code class="highlighter-rouge">scanpy</code>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</code></pre></div></div>

<p>A relatively high expression of mitochondria genes indicating a lower data quality in scRNA data. Therefore, we want to also remove cells with a high proportion of mitochondria genes expression.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="c1"># look for mitochondria genes which are started by 'MT'
</span><span class="n">mito_genes</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="nb">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">'MT-'</span><span class="p">)</span>
<span class="c1"># Get the mitochondria gene expression proportion
</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'percent_mito'</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="nb">sum</span><span class="p">(</span><span class="n">adata</span><span class="p">[:,</span> <span class="n">mito_genes</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="nb">sum</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">A1</span>
<span class="c1"># Keep only cell with a low mitochondria expression
</span><span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'percent_mito'</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">,</span> <span class="p">:]</span>
</code></pre></div></div>

<p>After filteration, we ahve 13714 cells left in the dataset.</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;&gt;&gt; adata
AnnData object with n_obs × n_vars = 2700 × 13714 
    obs: 'barcode', 'n_genes'
    var: 'gene_ids', 'gene_symbols', 'n_cells'
</code></pre></div></div>

<h2 id="22-normalization">2.2 Normalization</h2>
<p>The function <code class="highlighter-rouge">normalize_per_cell</code> normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default).</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">desc</span><span class="o">.</span><span class="n">normalize_per_cell</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">counts_per_cell_after</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
</code></pre></div></div>
<p>The figure shows the <strong>sorted</strong> totoal expression level for each cell before/after the normalization procedure.
<img src="assets/images/Figure_1_normalization.png" alt="desc normalization" /></p>

<h2 id="23-logarithm-transformation">2.3 logarithm transformation</h2>
<p>Also, the following analysis and the desc analysis should be performed on log-scaled data.</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">desc</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</code></pre></div></div>
<p>The figure shows a histogram of expression levels before/after logarithm transformation.
<img src="assets/images/Figure_2_log.png" alt="desc normalization" /></p>

<h2 id="24-selection-of-highly-variable-genes">2.4 Selection of highly variable genes</h2>
<p>The desc analysis should be performed on a particular set of genes, like highly variable genes, known cell-type markers or a combination of both, which help to separate cell groups. The function <code class="highlighter-rouge">highly_variable_genes</code> can help to select a set of the highly variable gene while controlling for the relationship between variability and average expression level. The function is also from <a href="https://scanpy.readthedocs.io/en/latest/">scanpy</a>. Check the <a href="https://icb-scanpy.readthedocs-hosted.com/en/latest/api/scanpy.api.pp.highly_variable_genes.html">function document</a> for detailed information about usage and parameter setting.</p>

<p>For reducing the compuation cost in the tutorial, we use <strong>top 100</strong> highly variable genes in the pbmc data for the following analysis:</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">n_top_genes</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">desc</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="n">n_top_genes</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</code></pre></div></div>

<h2 id="24-scaling">2.4 Scaling</h2>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">desc</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">zero_center</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</code></pre></div></div>
<p>The figure shows histogram of expreesion levels befter/after scaling.
<img src="assets/images/Figure_3_scaling.png" alt="desc normalization" /></p>

<p><br /></p>

<h1 id="3-desc-analysis">3 Desc analysis</h1>

<p>Now we are ready to run a desc analysis. The function <code class="highlighter-rouge">train</code> will perform desc on the expression matrix (2700 x 100 in this tutorial) and save the clustering labels as well other results in the AnnData object. For a full list of desc parameter please check the desc documentation on the <a href="https://pypi.org/project/desc/">pypi</a>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">dims</span><span class="o">=</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">n_top_genes</span><span class="p">),</span> <span class="mi">32</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span>
</code></pre></div></div>

<p>After the desc model training, several result slots will be added to the adata:</p>

<ul>
  <li><strong>adata.obs[‘ident’]</strong><br />
This slot is the clustering labels from the desc. The size of the ‘ident’ is <em>number of the cell by one</em> (2700 x 1 in this tutorial). 
One may want to check the number of cell for each of the clusters by:
    <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'ident'</span><span class="p">]):</span>
  <span class="k">print</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'ident'</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
</code></pre></div>    </div>
  </li>
  <li>
    <p><strong>adata.obsm[‘prob’].</strong><br />
This slot contains the probabilties of each cell belongs to each of the cell clusters. The size should be <em>numbers of cell by numbers of clusters</em> (2700 x ?? in this tutorial)</p>
  </li>
  <li><strong>adata.obs[‘max_prob’].</strong><br />
This slot contains the probabilities of each cell belongs to the assigned cluster. The size is <em>number of the cell by one</em> (2700 x 1 in this tutorial). A reliable assignment should base on a relatively high <strong>max_prob</strong>. Let us check <strong>max_prob</strong> for all cells using a scatter plot.</li>
</ul>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">max_prob</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s">'max_prob'</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">max_prob</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">max_prob</span><span class="p">))),</span> <span class="n">max_prob</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s">'sorted max_prob'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s">'cell index'</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s">'max_prob'</span><span class="p">)</span>
</code></pre></div></div>
<p><img src="assets/images/Figure_4_max_prob.png" alt="max_prob" /></p>

<ul>
  <li><strong>adata.obsm[‘embedded’]</strong><br />
This slot is a low dimension representation of the expression data. the size is <em>number of the cell by the number of network nodes in the lowest layer</em> (2700 x 8 in this tutorial). This slot facilitates users to generate tSNE plot or do other downstream analysis.</li>
</ul>

<p><br /></p>

<h1 id="4-visualization">4 Visualization</h1>

<p>Next, let us visualize the desc clustering using a tSNE plot:</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">tsne</span> <span class="o">=</span> <span class="n">desc</span><span class="o">.</span><span class="n">run_tsne</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="n">perplexity</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">learning_rate</span><span class="o">=</span><span class="mi">500</span><span class="p">)</span>
</code></pre></div></div>
<p><img src="assets/images/Figure_5_tsne.png" alt="max_prob" /> 
Note that the <code class="highlighter-rouge">tun_tsne</code> function in the desc package is a warpper fo the <a href="https://github.com/DmitryUlyanov/Multicore-TSNE"><code class="highlighter-rouge">MulticoreTSNE</code></a>. The tSNE visualization can be quite depend on the parameter settings. Thus, we suggest to run <code class="highlighter-rouge">run_tsne</code> under a grid of parameter settings to get a optimal visualization. To know more about tNSE paramter, please check <a href="https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html">here</a>.</p>

<p><br /></p>

<h1 id="5-save-result">5 Save result</h1>
<p>Lastly, let us save the desc result for the further analysis:</p>

<h2 id="51-save-to-a-h5ad-file">5.1 Save to a .h5ad file</h2>
<p>The AnnData object can be save to a .h5ad file for further analysis in Python enviroument.</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">'../result/desc_result.h5ad'</span><span class="p">)</span>
</code></pre></div></div>

<h2 id="52-save-to-csv-files">5.2 Save to .csv files</h2>
<p>Also, the desc result can be separatly save to .csv files, which can be easily accessed using R, c or other tools for future analysis.</p>
<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">desc</span><span class="o">.</span><span class="n">write_desc_result</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span> <span class="n">save_dir</span><span class="o">=</span><span class="s">'tmp_result'</span><span class="p">)</span>
</code></pre></div></div>
<p>The function will generate 3 files including <strong>cluster_ident.cvs</strong>, <strong>prob_matrix.cvs</strong> and <strong>embedded.cvs</strong>, which conresponding to <strong>adata.obs[‘ident’]</strong>, <strong>adata.obsm[‘prob’].</strong> and <strong>adata.obsm[‘embedded’]</strong>.</p>


      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/desc/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">DESC</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">DESC</li><li><a class="u-email" href="mailto:lyuyafei@gmail.com">lyuyafei@gmail.com</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://github.com/yafei611"><svg class="svg-icon"><use xlink:href="/desc/assets/minima-social-icons.svg#github"></use></svg> <span class="username">yafei611</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p></p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
